lib/executor/src/executors/dedupe.rs

```diff
@@ -11,66 +12,52 @@ pub struct SharedResponse {
     pub body: Bytes,
 }
 
-#[derive(Debug, Clone, Eq)]
-pub struct RequestFingerprint {
-    method: Method,
-    url: Uri,
-    /// BTreeMap to ensure case-insensitivity and consistent order for hashing
-    headers: BTreeMap<String, String>,
-    body: Vec<u8>,
-}
+pub fn request_fingerprint(
+    method: &Method,
+    url: &Uri,
+    req_headers: &HeaderMap,
+    upstream_headers: &NtexHeaderMap,
+    body_bytes: &[u8],
+    fingerprint_headers: &[String],
+) -> u64 {
+    let build_hasher = ABuildHasher::default();
+    let mut hasher = build_hasher.build_hasher();
+
+    // BTreeMap to ensure case-insensitivity and consistent order for hashing
+    let mut headers = BTreeMap::new();
+    if fingerprint_headers.is_empty() {
+        // fingerprint all headers
 
-impl RequestFingerprint {
-    pub fn new(
-        method: &Method,
-        url: &Uri,
-        req_headers: &HeaderMap,
-        body_bytes: &[u8],
-        fingerprint_headers: &[String],
-    ) -> Self {
-        let mut headers = BTreeMap::new();
-        if fingerprint_headers.is_empty() {
-            // fingerprint all headers
-            for (key, value) in req_headers.iter() {
+        for (key, value) in req_headers.iter() {
+            if let Ok(value_str) = value.to_str() {
+                headers.insert(key.as_str(), value_str);
+            }
+        }
+        for (key, value) in upstream_headers.iter() {
+            if let Ok(value_str) = value.to_str() {
+                headers.insert(key.as_str(), value_str);
```

Comment:
Why do we respect the gateway request headers in the fingerprint? 
If a subgraph request doesn't use anything from the gateway request, then they can be deduplicated safely. So I think we should only respect subgraph request's headers. A subgraph request that doesn't use anything from GW request can be shared with other GW request that doesn't have a subgraph request uses the GW request headers no?

Reply:
explain to me the efficiency of the fingerprint

Reply:
The right word is ` the efficiency of the deduplication` actually.
Let's say I have two requests sent to the GW with different auth headers, and they both have a subgraph request which are identical per `body` (`query`, `variables` and `extensions`) and `headers`(content-type etc but no auth header.
Then adding GW request headers will not deduplicate these identical inflight subgraph requests because you have GW request headers in the fingerprint. But if we only respect what's sent(subgraph request), then these will be deduped which is more efficient or I am fully missing the point.